[WPE][GTK] Sandbox should not allow mounting entire home directory

mcatanzaro opened the pull request:
#### 4b0c2332b9a6becb2a7b84058c42f02e2ac00c75
<pre>
[WPE][GTK] Sandbox should not allow mounting entire home directory
<a href="https://bugs.webkit.org/show_bug.cgi?id=250231">https://bugs.webkit.org/show_bug.cgi?id=250231</a>

Reviewed by Carlos Garcia Campos.

We should crash if the application tries to mount /home or $HOME in the
sandbox. We cannot prevents apps from mounting whatever they wish, but
if they want to be stupid they should have to try somewhat harder than
that.

* Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
(pathIsUserHomeDirectory):
(pathIsBlocked):

Canonical link: <a href="https://commits.webkit.org/259309@main">https://commits.webkit.org/259309@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/61a8887aab39aaa748b43b1a6b84cdd701b43bb5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102842 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11960 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35868 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112096 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12977 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2869 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95090 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109766 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108616 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9967 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93167 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37607 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/91812 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24698 "Found 2 new test failures: http/tests/security/clean-origin-css-exposed-resource-timing.html, http/tests/security/cross-origin-clean-css-resource-timing.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79348 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5413 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26113 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5561 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2570 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11580 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45603 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6419 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7306 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
```diff
@@ -1346,6 +1347,26 @@ void webkit_web_context_set_sandbox_enabled(WebKitWebContext* context, gboolean
+        return true;
+    }
+
+    return !strcmp(resolvedPath.get(), resolvedHomeDirectory.get());
```
I was going to comment this being case-sensitive is sometimes wrong. But it turns out GLib assumes this everywhere already so its just not supported.

Note that bubblewrap does go out if its way to handle case-insensitive filesystems.

mcatanzaro replied:
Updated.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
```diff
@@ -1346,6 +1347,26 @@ void webkit_web_context_set_sandbox_enabled(WebKitWebContext* context, gboolean
 
+static bool pathIsHomeDirectory(const char* path)
+{
+    GUniquePtr<char> resolvedPath(realpath(path, nullptr));
```
Is it a good idea to free with `g_free()` something allocated with system malloc? I guess `g_free()` always ends up calling free nowadays, but still. I think it's better to stack allocate a buffer of `PATH_MAX` and pass it a argument to realpath

mcatanzaro replied:
Nope. _Technically_ it's guaranteed to be safe because the documentation says g_malloc() will always use system malloc(), but it seems like a bad idea to rely on that. Good catch.

I think we can simply use std::unique_ptr here instead.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
```diff
@@ -1346,6 +1347,26 @@ void webkit_web_context_set_sandbox_enabled(WebKitWebContext* context, gboolean
+    if (!strcmp(resolvedPath.get(), "/home"))
+        return true;
+
+    GUniquePtr<char> resolvedHomeDirectory(realpath(g_get_home_dir(), nullptr));
```
And same here.

mcatanzaro committed:
[WPE][GTK] Sandbox should not allow mounting entire home directory
https://bugs.webkit.org/show_bug.cgi?id=250231

Reviewed by Carlos Garcia Campos.

We should crash if the application tries to mount /home or $HOME in the
sandbox. We cannot prevents apps from mounting whatever they wish, but
if they want to be stupid they should have to try somewhat harder than
that.

* Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
(pathIsUserHomeDirectory):
(pathIsBlocked):

Canonical link: https://commits.webkit.org/259309@main
